samoa-api/src/main/java/org/apache/samoa/streams/kafka/KafkaReader.java

```diff
@@ -0,0 +1,279 @@
+        kafka.javaapi.OffsetRequest request = new kafka.javaapi.OffsetRequest(
+                requestInfo, kafka.api.OffsetRequest.CurrentVersion(), clientName);
+        OffsetResponse response = consumer.getOffsetsBefore(request);
+
+        if (response.hasError()) {
+            System.out.println("Error fetching data Offset Data the Broker. Reason: " + response.errorCode(topic, partition));
+            return 0;
+        }
+        long[] offsets = response.offsets(topic, partition);
+        return offsets[0];
+    }
+
+    /**
+     * Uses the findLeader() logic we defined to find the new leader, except here we only try to connect to one of the
+     * replicas for the topic/partition. This way if we can’t reach any of the Brokers with the data we are interested
+     * in we give up and exit hard.
+     *
+     * @param a_oldLeader
+     * @param a_topic
+     * @param a_partition
+     * @param a_port
+     * @return
+     * @throws Exception
+     */
+    private String findNewLeader(String a_oldLeader, String a_topic, int a_partition, int a_port) throws Exception {
+        for (int i = 0; i < 3; i++) {
+            boolean goToSleep = false;
+            PartitionMetadata metadata = findLeader(m_replicaBrokers, a_port, a_topic, a_partition);
+            if (metadata == null) {
+                goToSleep = true;
+            } else if (metadata.leader() == null) {
+                goToSleep = true;
+            } else if (a_oldLeader.equalsIgnoreCase(metadata.leader().host()) && i == 0) {
+                // first time through if the leader hasn't changed give ZooKeeper a second to recover
+                // second time, assume the broker did recover before failover, or it was a non-Broker issue
+                //
+                goToSleep = true;
+            } else {
+                return metadata.leader().host();
+            }
+            if (goToSleep) {
+                try {
+                    Thread.sleep(1000);
+                } catch (InterruptedException ie) {
+                }
+            }
+        }
+        System.out.println("Unable to find new leader after Broker failure. Exiting");
+        throw new Exception("Unable to find new leader after Broker failure. Exiting");
+    }
+
+    /**
+     * Query a live broker to find out leader information and replica information for a given topic and partition
+     *
+     * @param a_seedBrokers
+     * @param a_port
+     * @param a_topic
+     * @param a_partition
+     * @return
+     */
+    private PartitionMetadata findLeader(List<String> a_seedBrokers, int a_port, String a_topic, int a_partition) {
+        PartitionMetadata returnMetaData = null;
+        for (String seed : a_seedBrokers) {
+            SimpleConsumer consumer = null;
+            try {
+                consumer = new SimpleConsumer(seed, a_port, 100000, 64 * 1024, "leaderLookup"); //broker_host, broker_port, timeout, buffer_size, client_id
```

Comment:
I am not fully aware of Kafka internals, but I know that they have two APIs for consumers. Does this one use the new API?